File: ResourceManager/ViewModels/Tools/DatabaseContentViewModel.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Microsoft.Win32;
using NetCore.DatabaseToolkit.SQLite;
using NetCore.Mvvm.Commands;
using NetCore.Mvvm.Controls;
using NetCore.Validators;
using ResourceManager.Configuration;
using ResourceManager.Helpers;
using ResourceManager.Settings;
using System;
using System.Windows.Input;

namespace ResourceManager.ViewModels
{
    public class DatabaseContentViewModel : PageContentViewModel
    {
        private AppSettings appSettings => App.ServiceProvider.GetRequiredService<IOptions<AppSettings>>().Value;
        public DatabaseContentViewModel()
        {
            MenuTitle = "Database";
            MenuIcon = AssetManager.Instance.GetImage("Database.png");
            Title = "Database Tools";
            PageColor = Constants.ToolsPageColor;

            AddValidationRule(() => FilePath, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(FilePath)));

            ExecuteCommand = new RelayCommand(OnExecute, CanExecute);
            OpenCommand = new RelayCommand(OnOpen);

            IsBackup = true;
        }

        private void OnOpen()
        {
            if (IsBackup)
            {
                var dlg = new SaveFileDialog();
                dlg.DefaultExt = ".bak";
                dlg.Filter = "Backup File|*.bak";
                dlg.Title = "Save file to backup";
                var b = dlg.ShowDialog();
                if (b != null & b.Equals(true))
                {
                    FilePath = dlg.FileName;
                }
            }
            else
            {
                var dlg = new OpenFileDialog();
                dlg.DefaultExt = ".bak";
                dlg.Filter = "Backup File|*.bak";
                dlg.Title = "Open file to restore";
                var b = dlg.ShowDialog();
                if (b != null & b.Equals(true))
                {
                    FilePath = dlg.FileName;
                }
            }
        }

        private bool CanExecute()
        {
            return !HasErrors;
        }

        private void OnExecute()
        {
            ValidateAllRules();
            if (!HasErrors)
            {
                try
                {
                    ISQLiteToolkit tools = App.ServiceProvider.GetRequiredService<ISQLiteToolkit>();
                    string dbFilename = Constants.DbFileName;
                    if (IsBackup)
                    {
                        int code;
                        string output;
                        (code, output) = tools.BackupDatabase(dbFilename, FilePath);
                        if (code == 0)
                        {
                            ShowPopupMessage("Database backed up succesfully.", appSettings.AppName, PopupButton.OK, PopupImage.Information);
                        }
                        else
                        {
                            ShowPopupMessage("Backup database failed. Message:" + output, appSettings.AppName, PopupButton.OK, PopupImage.Warning);
                        }
                    }
                    else
                    {
                        int code;
                        string output;
                        (code, output) = tools.RestoreDatabase(dbFilename, FilePath);
                        if (code == 0)
                        {
                            ShowPopupMessage("Database restored succesfully.", appSettings.AppName, PopupButton.OK, PopupImage.Information);
                        }
                        else
                        {
                            ShowPopupMessage("Restore database failed. Message:" + output, appSettings.AppName, PopupButton.OK, PopupImage.Warning);
                        }
                    }
                }
                catch (Exception e)
                {
                    windowManager.ShowMessageBox(e.Message, System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
                }
            }
        }

        public bool IsBackup
        {
            get => GetProperty<bool>();
            set => SetProperty(value);
        }

        public string FilePath
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }

        public ICommand ExecuteCommand { get; }

        public ICommand OpenCommand { get; }
    }
}

File: ResourceManager/ViewModels/Tools/DatabaseContentViewModel.vb
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Extensions.Options
Imports Microsoft.Win32
Imports VBNetCore.DatabaseToolkit.SQLite
Imports VBNetCore.Mvvm.Commands
Imports VBNetCore.Mvvm.Controls
Imports VBNetCore.Validators
Imports ResourceManager.Configuration
Imports ResourceManager.Helpers
Imports ResourceManager.Settings
Imports System
Imports System.Windows.Input
Imports SQLitePCL

Namespace ResourceManager.ViewModels
    Public Class DatabaseContentViewModel
        Inherits PageContentViewModel
        Private ReadOnly Property appSettings As AppSettings
            Get
                Return ServiceProviderServiceExtensions.GetRequiredService(Of IOptions(Of AppSettings))(Application.ServiceProvider).Value
            End Get
        End Property
        Public Sub New()
            MenuTitle = "Database"
            MenuIcon = AssetManager.Instance.GetImage("Database.png")
            Title = "Database Tools"
            PageColor = ToolsPageColor

            AddValidationRule(Function() FilePath, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(DatabaseContentViewModel.FilePath)))

            ExecuteCommand = New RelayCommand(AddressOf OnExecute, AddressOf CanExecute)
            OpenCommand = New RelayCommand(AddressOf OnOpen)

            IsBackup = True
        End Sub

        Private Sub OnOpen()
            If IsBackup Then
                Dim dlg = New SaveFileDialog()
                dlg.DefaultExt = ".bak"
                dlg.Filter = "Backup File|*.bak"
                dlg.Title = "Save file to backup"
                Dim b = dlg.ShowDialog()
                If b IsNot Nothing And b.Equals(True) Then
                    FilePath = dlg.FileName
                End If
            Else
                Dim dlg = New OpenFileDialog()
                dlg.DefaultExt = ".bak"
                dlg.Filter = "Backup File|*.bak"
                dlg.Title = "Open file to restore"
                Dim b = dlg.ShowDialog()
                If b IsNot Nothing And b.Equals(True) Then
                    FilePath = dlg.FileName
                End If
            End If
        End Sub

        Private Function CanExecute() As Boolean
            Return Not HasErrors
        End Function

        Private Sub OnExecute()
            ValidateAllRules()
            If Not HasErrors Then
                Try
                    Dim tools As ISQLiteToolkit = Application.ServiceProvider.GetRequiredService(Of ISQLiteToolkit)()
                    Dim dbFilename = Constants.DbFileName
                    If IsBackup Then
                        Dim x = tools.BackupDatabase(dbFilename, FilePath)
                        If x.Item1 = 0 Then
                            ShowPopupMessage("Database backed up succesfully.", appSettings.AppName, PopupButton.OK, PopupImage.Information)
                        Else
                            ShowPopupMessage("Backup database failed. Message:" & x.Item2, appSettings.AppName, PopupButton.OK, PopupImage.Warning)
                        End If
                    Else
                        Dim x = tools.RestoreDatabase(dbFilename, FilePath)
                        If x.Item1 = 0 Then
                            ShowPopupMessage("Database restored succesfully.", appSettings.AppName, PopupButton.OK, PopupImage.Information)
                        Else
                            ShowPopupMessage("Restore database failed. Message:" & x.Item2, appSettings.AppName, PopupButton.OK, PopupImage.Warning)
                        End If
                    End If
                Catch e As Exception
                    windowManager.ShowMessageBox(e.Message, System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error)
                End Try
            End If
        End Sub

        Public Property IsBackup As Boolean
            Get
                Return GetProperty(Of Boolean)()
            End Get
            Set(ByVal value As Boolean)
                SetProperty(value)
            End Set
        End Property

        Public Property FilePath As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property

        Public ReadOnly Property ExecuteCommand As ICommand
        Public ReadOnly Property OpenCommand As ICommand
    End Class
End Namespace
